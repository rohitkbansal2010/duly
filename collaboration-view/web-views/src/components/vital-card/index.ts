export { VitalCard } from './vital-card';
export { VitalCardSkeleton } from './vital-card-skeleton';
export { vitalsCardsMap } from './vitals-cards-map';
