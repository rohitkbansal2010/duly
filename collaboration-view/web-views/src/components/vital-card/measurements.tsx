import { measurementsUnitsMap } from '@constants';
import { MeasurementType } from '@enums';
import { Measurement } from '@types';

import styles from './vital-card.scss';

type MeasurementsTypeProps = {
  index: number;
  measurements: Measurement[];
  activeClass: string;
  isWeighHeightCard: boolean;
  isVitalHeight: boolean;
  isVitalWeight: boolean;
};

export const Measurements = ({
  index,
  measurements,
  activeClass,
  isWeighHeightCard,
  isVitalHeight,
  isVitalWeight,
}:
  MeasurementsTypeProps) => {
  const measurementsTypesWeightAndHeight = [
    MeasurementType.BODY_WEIGHT,
    MeasurementType.BODY_HEIGHT,
  ];

  const renderMeasurement = ({
    measurementType,
    value,
    unit,
    convertedValue,
    convertedUnit,
    maxScaleValue,
  }: Measurement) => {
    const metric = sessionStorage.getItem('metric') === 'on';
    const measurementValueClasses = [ styles[`todayVitalCard${activeClass}MeasurementValue`] ]
      .concat(
        measurementsTypesWeightAndHeight.includes(measurementType)
          ? styles[`todayVitalCard${activeClass}MeasurementValueWeightHeight`]
          : []
      )
      .join(' ');

    const isSiastolicBloodPressure = measurementType === MeasurementType.SYSTOLIC_BLOOD_PRESSURE;
    const isDiastolicBloodPressure = measurementType === MeasurementType.DIASTOLIC_BLOOD_PRESSURE;
    const isPainLevel = measurementType === MeasurementType.PAIN_LEVEL;

    return (
      <span key={measurementType} className={styles[`todayVitalCard${activeClass}Measurement`]}>
        {!isDiastolicBloodPressure && (
          <span className={measurementValueClasses}>
            {!metric && convertedValue ? convertedValue : value}
            {!(isSiastolicBloodPressure || maxScaleValue) && ' '}
          </span>
        )}
        {unit && !isPainLevel && (
          <span className={styles[`todayVitalCard${activeClass}MeasurementUnit`]}>
            {isDiastolicBloodPressure && (
              <>&nbsp;/&nbsp;{!metric && convertedValue ? convertedValue : value}&nbsp;</>
            )}
            {
            measurementsUnitsMap[measurementType][(!metric && convertedUnit) ? convertedUnit : unit]
            }
          </span>
        )}
        {maxScaleValue && (
          <span className={styles[`todayVitalCard${activeClass}MeasurementScaleValue`]}>
            &nbsp;/&nbsp;{maxScaleValue}
          </span>
        )}
      </span>
    );
  };

  return (
    <span className={styles[`todayVitalCard${activeClass}Measurement`]}>
      {measurements.map(renderMeasurement)}
      {isWeighHeightCard && isVitalHeight && isVitalWeight && index == 0 && <>&nbsp;/&nbsp;</>}
    </span>
  );
};
