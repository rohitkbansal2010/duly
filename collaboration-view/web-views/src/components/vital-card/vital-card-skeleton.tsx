import { DulyLoader } from '@components/duly-loader';
import { VITALS_LOADER_WIDTH } from '@constants';
import { vitalSkeletonIcon } from '@icons';

import styles from './vital-card.scss';

export const VitalCardSkeleton = () =>
  (
    <div className={styles.todayVitalCardSkeleton}>
      <div className={styles.todayVitalCardSkeletonImage} >
        <img src={vitalSkeletonIcon} alt="vital loading" />
      </div>
      <div className={styles.todayVitalCardSkeletonLoader}>
        <DulyLoader width={VITALS_LOADER_WIDTH} />
      </div>
    </div>
  );
