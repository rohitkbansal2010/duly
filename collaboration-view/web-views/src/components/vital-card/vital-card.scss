@import "~@styles/colors.scss";
@import "~@styles/typography.scss";
@import "~@styles/main.scss";

.todayVitalCard {
  width: calc(50% - 0.75rem);
  height: 10.4rem;
  padding: 1.125rem 0 0.875rem 1.25rem;
  text-align: left;
  background-color: $white;
  border: none;
  border-radius: 0.5rem;
  box-shadow: 0 0.25rem 1.875rem rgba(0, 178, 240, 0.1);

  &:last-child,
  &:nth-child(7) {
    margin-bottom: 0;
  }

  &Icon {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 3.75rem;
    height: 3.75rem;
    margin-bottom: 0.75rem;
    border-radius: 50%;
    @include background-opacity($dupage-light-blue, 0.3);
  }

  &Icon,
  &ActiveIcon {
    & img {
      object-fit: cover;
    }

    &BloodPressure {
      width: 2.4rem;
      height: 2.02rem;
    }

    &BloodOxygen {
      width: 1.66rem;
      height: 2.75rem;
    }

    &HeartRate {
      width: 2.29rem;
      height: 2.1rem;
    }

    &Temperature {
      width: 1.38rem;
      height: 2.5rem;
    }

    &RespiratoryRate {
      width: 2.29rem;
      height: 2.29rem;
    }

    &PainLevel {
      width: 1.16rem;
      height: 2.06rem;
    }

    &WeightAndHeight {
      width: 2.29rem;
      height: 2.29rem;
    }

    &BodyMassIndex {
      width: 2.29rem;
      height: 2.29rem;
    }
  }

  &Title {
    @include inter-text-default-bold;
    margin-bottom: 0.25rem;
    color: $dupage-dark-blue;
    font-size: var(--font-size-16px-4px);
  }

  &Measurement,
  &ActiveMeasurement {
    color: $dupage-dark-blue;

    &ValueEmpty,
    &Value {
      @include nexa-heavy-highlight;
      font-size: 2rem;
      line-height: 1.375rem;
    }

    &ScaleValue,
    &Unit {
      @include nexa-heavy-highlight;
      font-size: 1.25rem;
      line-height: 1.25rem;
    }

    &ValueWeightHeight {
      @include nexa-heavy-highlight;
      font-size: 1.5rem;
      line-height: 1.25rem;
    }
  }

  &Active {
    @extend .todayVitalCard;
    background-color: $dupage-dark-blue;

    &Icon {
      @extend .todayVitalCardIcon;
      @include background-opacity($white, 0.7);
    }

    &Title {
      @extend .todayVitalCardTitle;
      color: $white;
      font-size: var(--font-size-16px-4px) !important;
    }

    &Measurement,
    &MeasurementValueEmpty {
      color: $white;
    }
  }

  &Skeleton {
    display: flex;
    flex-direction: column;
    width: calc(50% - 0.75rem);
    height: 10.4rem;
    background: $white;
    border-radius: 0.5rem;

    &Image {
      display: flex;
      flex-shrink: 0;
      align-items: center;
      justify-content: center;
      width: 3.75rem;
      height: 3.75rem;
      margin-top: 1.25rem;
      margin-left: 1.25rem;
      background: rgba(0, 165, 223, 0.3);
      border-radius: 50%;

      img {
        width: 2rem;
      }
    }

    &Loader {
      position: relative;
      width: max-content;
      height: 100%;
      margin-top: 1.25rem;
    }
  }
}
