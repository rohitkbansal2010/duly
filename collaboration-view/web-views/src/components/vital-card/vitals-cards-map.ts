import { CardType } from '@enums';
import {
  conditionLightBlueIcon,
  conditionIcon,
  bloodPressureLightBlueIcon,
  bloodPressureIcon,
  heartRateLightBlueIcon,
  heartRateIcon,
  temperatureLightBlueIcon,
  temperatureIcon,
  respiratoryRateLightBlueIcon,
  respiratoryRateIcon,
  painLevelLightBlueIcon,
  painLevelIcon,
  weightLightBlueIcon,
  weightIcon,
  bmiLightBlueIcon,
  bmiIcon
} from '@icons';

type VitalsCardsMapType = {
  [key in CardType]: {
    title: string;
    icon: string;
    activeIcon: string;
  };
};

export const vitalsCardsMap: VitalsCardsMapType = {
  [CardType.BLOOD_PRESSURE]: {
    title: 'BLOOD PRESSURE',
    icon: conditionLightBlueIcon,
    activeIcon: conditionIcon,
  },
  [CardType.BLOOD_OXYGEN]: {
    title: 'BLOOD OXYGEN',
    icon: bloodPressureLightBlueIcon,
    activeIcon: bloodPressureIcon,
  },
  [CardType.HEART_RATE]: {
    title: 'HEART RATE',
    icon: heartRateLightBlueIcon,
    activeIcon: heartRateIcon,
  },
  [CardType.TEMPERATURE]: {
    title: 'TEMPERATURE',
    icon: temperatureLightBlueIcon,
    activeIcon: temperatureIcon,
  },
  [CardType.RESPIRATORY_RATE]: {
    title: 'RESPIRATORY RATE',
    icon: respiratoryRateLightBlueIcon,
    activeIcon: respiratoryRateIcon,
  },
  [CardType.PAIN_LEVEL]: {
    title: 'PAIN LEVEL',
    icon: painLevelLightBlueIcon,
    activeIcon: painLevelIcon,
  },
  [CardType.WEIGHT_HEIGHT]: {
    title: 'WEIGHT & HEIGHT',
    icon: weightLightBlueIcon,
    activeIcon: weightIcon,
  },
  [CardType.BODY_MASS_INDEX]: {
    title: 'BMI',
    icon: bmiLightBlueIcon,
    activeIcon: bmiIcon,
  },
};
