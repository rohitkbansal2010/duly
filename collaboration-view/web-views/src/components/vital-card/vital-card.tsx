import { CardType, VitalType } from '@enums';
import { TodayVitalCard } from '@types';
import { getTitleByCardType } from '@utils';

import { Measurements } from './measurements';
import styles from './vital-card.scss';
import { vitalsCardsMap } from './vitals-cards-map';

export type VitalCardPropType = TodayVitalCard & {
  isActiveCard: boolean;
  onClickVitalCard: () => void;
}

export const VitalCard = ({
  cardType,
  vitals,
  isActiveCard,
  onClickVitalCard,
}: VitalCardPropType) => {
  const activeClass = isActiveCard ? 'Active' : '';

  const { icon, activeIcon } = vitalsCardsMap[cardType];

  const isWeighHeightCard = cardType.includes(CardType.WEIGHT_HEIGHT);
  const isBMICard = cardType.includes(CardType.BODY_MASS_INDEX);

  const isVitalHeightNoData =
    isWeighHeightCard &&
    !vitals[1] &&
    vitals[0]?.vitalType.includes(VitalType.WEIGHT);

  const isVitalWeightNoData =
    isWeighHeightCard &&
    !vitals[1] &&
    vitals[0]?.vitalType.includes(VitalType.HEIGHT);

  return (
    <button
      onClick={onClickVitalCard}
      data-testid="vital-card__container"
      className={styles[`todayVitalCard${activeClass}`]}
    >
      <div className={styles[`todayVitalCard${activeClass}Icon`]}>
        <img
          data-testid="vital-card__icon"
          src={isActiveCard ? activeIcon : icon}
          alt="vital icon"
          className={styles[`todayVitalCard${activeClass}Icon${cardType}`]}
        />
      </div>
      <div data-testid="vital-card__title" className={styles[`todayVitalCard${activeClass}Title`]}>
        {getTitleByCardType(vitals, cardType)}
      </div>
      {!vitals.length && (
        <span
          data-testid="vital-card__no-data"
          className={styles[`todayVitalCard${activeClass}MeasurementValueEmpty`]}
        >
          N/A
        </span>
      )}
      {isVitalWeightNoData && (
        <span
          data-testid="vital-card__weight__no-data"
          className={styles[`todayVitalCard${activeClass}MeasurementValueEmpty`]}
        >
          N/A&nbsp;/&nbsp;
        </span>
      )}
      {vitals
        .slice(0, !isBMICard ? vitals.length : 1)
        .map(({ vitalType, measurements }, index) =>
          (
            <Measurements
              key={vitalType}
              index={index}
              measurements={measurements}
              activeClass={activeClass}
              isWeighHeightCard={isWeighHeightCard}
              isVitalHeight={!isVitalHeightNoData}
              isVitalWeight={!isVitalWeightNoData}
              
              
            />
          ))}
      {isVitalHeightNoData && (
        <span
          data-testid="vital-card__height__no-data"
          className={styles[`todayVitalCard${activeClass}MeasurementValueEmpty`]}
        >
          &nbsp;/&nbsp;N/A
        </span>
      )}
    </button>
  );
};
