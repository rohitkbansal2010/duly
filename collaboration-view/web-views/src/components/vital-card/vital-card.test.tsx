import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Measurements } from '@components/vital-card/measurements';
import { VitalCard, VitalCardPropType } from '@components/vital-card/vital-card';
import { vitalsCardsMap } from '@components/vital-card/vitals-cards-map';
import { CardType, MeasurementType, VitalType } from '@enums';

describe('VitalCard', () => {
  let Component: ReturnType<typeof shallow>;
  const cardType = CardType.HEART_RATE;
  const props: VitalCardPropType = {
    isActiveCard: false,
    onClickVitalCard: jest.fn(),
    cardType,
    vitals: [
      {
        vitalType: VitalType.HEART_RATE,
        measurements: [
          {
            measurementType: MeasurementType.HEART_RATE,
            value: 1,
            measured: '',
          },
        ],
      },
      {
        vitalType: VitalType.PAIN_LEVEL,
        measurements: [
          {
            measurementType: MeasurementType.PAIN_LEVEL,
            value: 1,
            measured: '',
          },
        ],
      },
    ],
  };

  beforeEach(() => {
    Component = shallow(<VitalCard {...props} />);
  });

  it('should render correctly', () => {
    expect(toJson(Component)).toMatchSnapshot();
  });

  it('should handle click on vital card', () => {
    const container = Component.find({ 'data-testid': 'vital-card__container' });
    container.simulate('click');

    expect(props.onClickVitalCard).toHaveBeenCalled();
  });

  it('should render measurements for each vital', () => {
    expect(Component.find(Measurements)).toHaveLength(props.vitals.length);
  });

  it('should render no data text if there are no vitals', () => {
    const wrapper = shallow(<VitalCard {...props} vitals={[]}/>);

    expect(wrapper.find({ 'data-testid': 'vital-card__no-data' }).exists()).toBe(true);
    expect(wrapper.find(Measurements)).toHaveLength(0);
  });

  it('should render only one measurement if card type is BMI', () => {
    const wrapper = shallow(<VitalCard {...props} cardType={CardType.BODY_MASS_INDEX} />);
    expect(wrapper.find(Measurements)).toHaveLength(1);
  });

  it('should render vital height no data text', () => {
    const vitals = [
      {
        vitalType: VitalType.WEIGHT,
        measurements: [],
      },
    ];

    const wrapper = shallow(
      <VitalCard
        {...props}
        cardType={CardType.WEIGHT_HEIGHT}
        vitals={vitals}
      />
    );

    expect(wrapper.find({ 'data-testid': 'vital-card__height__no-data' }).exists()).toBe(true);
  });

  it('should render vital weight no data text', () => {
    const vitals = [
      {
        vitalType: VitalType.HEIGHT,
        measurements: [],
      },
    ];

    const wrapper = shallow(
      <VitalCard
        {...props}
        cardType={CardType.WEIGHT_HEIGHT}
        vitals={vitals}
      />
    );

    expect(wrapper.find({ 'data-testid': 'vital-card__weight__no-data' }).exists()).toBe(true);
  });

  it('should render normal icon if vital card is not active', () => {
    const vitalCardData = vitalsCardsMap[cardType];
    expect(Component.find({ 'data-testid': 'vital-card__icon' }).props().src).toBe(vitalCardData.icon);
  });

  it('should render active icon if vital card is active', () => {
    const wrapper = shallow(<VitalCard {...props} isActiveCard />);
    const vitalCardData = vitalsCardsMap[cardType];

    expect(wrapper.find({ 'data-testid': 'vital-card__icon' }).props().src).toBe(vitalCardData.activeIcon);
  });
});

describe('vitalCardsMap', () => {
  const vitalCardTypes = Object.keys(vitalsCardsMap);
  test.each(vitalCardTypes)('should render correctly "%s" vital card', (vitalCardType) => {
    const props: VitalCardPropType = {
      isActiveCard: false,
      onClickVitalCard: jest.fn(),
      cardType: CardType.HEART_RATE,
      vitals: [],
    };
    const wrapper = shallow(<VitalCard {...props} cardType={vitalCardType as CardType} />);

    const vitalCardData = vitalsCardsMap[vitalCardType as CardType];
    expect(wrapper.find({ 'data-testid': 'vital-card__title' }).text()).toBe(vitalCardData.title);
    expect(wrapper.find({ 'data-testid': 'vital-card__icon' }).props().src).toBe(vitalCardData.icon);
  });
});
