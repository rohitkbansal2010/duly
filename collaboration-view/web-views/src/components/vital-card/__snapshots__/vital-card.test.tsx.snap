// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VitalCard should render correctly 1`] = `
<button
  className="todayVitalCard"
  data-testid="vital-card__container"
  onClick={[MockFunction]}
>
  <div
    className="todayVitalCardIcon"
  >
    <img
      alt="vital icon"
      className="todayVitalCardIconHeartRate"
      data-testid="vital-card__icon"
      src="heart-rate-light-blue.svg"
    />
  </div>
  <div
    className="todayVitalCardTitle"
    data-testid="vital-card__title"
  >
    HEART RATE
  </div>
  <Measurements
    activeClass=""
    index={0}
    isVitalHeight={true}
    isVitalWeight={true}
    isWeighHeightCard={false}
    key="HeartRate"
    measurements={
      Array [
        Object {
          "measured": "",
          "measurementType": "HeartRate",
          "value": 1,
        },
      ]
    }
  />
  <Measurements
    activeClass=""
    index={1}
    isVitalHeight={true}
    isVitalWeight={true}
    isWeighHeightCard={false}
    key="PainLevel"
    measurements={
      Array [
        Object {
          "measured": "",
          "measurementType": "PainLevel",
          "value": 1,
        },
      ]
    }
  />
</button>
`;
